### ISSUE TITLE
Ethers.js MetaMask Mobile error: eth_sendTransaction

### ISSUE THREAD

**Issue by pi0neerpat** [WalletConnect/walletconnect-monorepo#309]
Not sure where this issue originates - MetaMask, Ethers.js, or here. Apologies if this is the wrong place.

### Issue

Attempting to fire a transaction fails with error `The method eth_sendTransaction does not exist/is not available`.

This flow works for WalletConnect with the test wallet (test.walletconnect.org). It also works well with MetaMask browser extension (ie. not using WalletConnect)

`@walletconnect/web3-provider` v1.0.13
`@ethersproject/providers` ^5.0.4

```js
const provider = new WalletConnectProvider({
      infuraId: infuraKey // Required
});

await provider.enable();
const walletProvider = new Web3Provider(provider);

contracts[name] = new Contract(
   CONTRACTS[name][networkName],
    CONTRACTS[name].abi,
    walletProvider
);

const approvalAmount = amountBn || parseUnits("500000", 18);
const tx = await contracts[token].approve(
  spender,
  approvalAmount
);
```

### Discussion

- It looks like ethers.js supports eth_sendTransaction https://github.com/ethers-io/ethers.js/issues/298
- It looks like MetaMask Mobile supports eth_sendTransaction https://github.com/MetaMask/metamask-mobile/blob/aa2895d5c08469c8a23522b455bed55acc8d11a1/app/core/WalletConnect.js#L90
- It looks like WalletConnect should support ethers providers and MetaMask Mobile https://github.com/ethers-io/ethers.js/issues/775

Going to continue researching....

**Comment by pedrouid** [comment]
Hey @pi0neerpat I would love to know anything you find because I’ve been banging my head against the wall for over a month to solve this

I have created examples both on the monorepo and a simple create-react-app to reproduce this and it’s consistently failing.

I’ve got multiple reports from this issue and I’ve also communicated with @ricmoo about it too. 

Anything you find would be extremely helpful 🙌

**Comment by pi0neerpat** [comment]
Ok thanks for confirming this isn't a "just me" issue. I will do some poking.

Do you have links to these multiple reports? I didn't find any other issues that might help me

**Comment by pedrouid** [comment]
I mean I have screenshots of the exact issue you reported.

This one happens when you pass the Web3Provider to the Contract class
<img width="677" alt="Screenshot 2020-07-02 19 48 39" src="https://user-images.githubusercontent.com/10136079/87167783-cc4e7700-c2cd-11ea-9e6b-9e3281816d5a.png">

This one happens when you pass the Web3Provider.JsonRpcSigner to the Contract class
<img width="989" alt="Screenshot 2020-07-02 18 13 49" src="https://user-images.githubusercontent.com/10136079/87167780-c9ec1d00-c2cd-11ea-814b-ecfec0bfdf30.png">

Also created this example that tests both a DAI transfer and a personal_sign message
https://github.com/pedrouid/walletconnect-ethers-example

This is where it left me really confused because it has no issues signing the message but it fails to send the transaction

I've looked into the ethers.js source code to identify where this error is triggered:
https://github.com/ethers-io/ethers.js/blob/master/packages/contracts/src.ts/index.ts#L311

Contract has a property for signer which is defined by the check `Signer.isSigner()` on the constructor. This check isSigner checks if the provider has a truthy readonly property called `_isSigner`:
https://github.com/ethers-io/ethers.js/blob/master/packages/abstract-signer/src.ts/index.ts#L206

Whenever WalletConnect provider is used with ethers.js is wrapped with the providers.Web3Provider which extends from providers.JsonRpcProvider

JsonRpcProvider includes a JsonRpcSigner which is where I start loosing track at where the issue might occur because I'm not entirely sure what WalletConnect provider need to do to succeed the Signer.isSigner check

PS - I've tried adding `_isSigner = true` to WalletConnect provider but that didn't solve it

**Comment by mds1** [comment]
Hey, just wanted to see what the latest is on this issue and if there's any known workarounds? Thought I must have been making a simple mistake but apparently this is quite the tricky issue. In this instance I'm seeing the error when trying to send a transaction with ethers + Argent via WalletConnect

- WalletConnect v1.1.0
- Web3Modal v1.9.0
- ethers v5.0.7

**Comment by BSathvik** [comment]
I'm having the same issue as well when trying to sign using `eth_signTypedData` on MetaMask mobile:
- `web3.js` v1.2.11
- `@walletconnect/web3-provider` v1.1.0

**Comment by moodysalem** [comment]
We need to upgrade `@ethersproject/contract` in the Uniswap Interface, due to https://github.com/ethers-io/ethers.js/issues/847
This issue is blocking since the upgrade breaks WalletConnect due to this error
@ricmoo do you have any insights?

**Comment by ricmoo** [comment]
Oh, sorry. What is the problem with using value in the overrides?

**Comment by moodysalem** [comment]
> Oh, sorry. What is the problem with using value in the overrides?

In the linked issue, if you pass value to overrides for `contract[methodName].staticCall`, it throws

I want to do this in the interface to get the revert reason for a simulated swap and show a good error message if it fails
https://github.com/Uniswap/uniswap-interface/pull/1015/commits/b67d5aa8b6b48335d710be018edba5e3b45dde93#diff-54ce46b532a6309e705bf4a28a50f8aeR161

**Comment by ricmoo** [comment]
@moodysalem Does this still happen with the latest version? I thought this was fixed? What version are you using? Can you provide a code snippet with an ABI?

**Comment by moodysalem** [comment]
> @moodysalem Does this still happen with the latest version? I thought this was fixed? What version are you using? Can you provide a code snippet with an ABI?

No. The ethers issue I linked is fixed in the latest version.

However, this _WalletConnect_ issue on which we are commenting is broken with the latest ethers version. So we cannot upgrade to the latest ethers version in the Uniswap Interface without breaking WalletConnect. And we need to upgrade ethers to fix the ethers issue.

I'm asking if you have any insight into how to fix this WalletConnect/ethers compatibility issue in the latest version of ethers.

**Comment by ricmoo** [comment]
Oh! I see. :)

I just re-read the above, and it looks like you are trying to use a Provider as a Signer? If you pass a provider into a contract, the contract is read-only. You must pass in a Signer.

I haven't had a chance to get into WalletConnect yet, but there should be something like:

```javascript
class WalltConnectSigner extends ethers.Signer {
  constructor(...) {
    super();
    // Other setup here
  }
}
```

If you want to be able to execute write operations, you must pass in an object that extends Signer. If you are creating a speical provider, you should override the `.getSigner` method on it, to wrap the `super.getSigner()` with whatever extras the WalletConnectSigner will need.

My guess though, is that you don't even need a WalletConnectProvider, but can create just a WalletConnectSigner class. Keep in mind ethers creates a very strong distinction between the two: Providers are read-only blockchain state; Signers have the ability to sign, or something to that extent.

If you look at the GsnSigner (there is a PR open by @yuetloo) on ethers, you will see she only needed to create a Signer sub-class.

Does that make sense?

**Comment by moodysalem** [comment]
@pedrouid it seems like when `eth_sendTransaction` is called on the walletconnect web3-provider, it goes to the infura node because of this line
https://github.com/WalletConnect/walletconnect-monorepo/blob/916621262d44989e905c1f667a1d567e7982ba22/packages/providers/web3-provider/src/index.ts#L277

I found this behavior without ethers, by constructing a WalletConnect web3 provider and calling `provider.send({method:'eth_sendTransaction', params: []})`

EDIT: Actually I just noticed this issue, which looks like the same issue
https://github.com/WalletConnect/walletconnect-monorepo/issues/319

**Comment by waterdrop01** [comment]
I have the exact same issue when injecting WalletConnect into Ethers.js.

Any workaround to make Ethers.js send the signing request to WalletConnect instead of Infura would be awesome!

![Screenshot from 2020-08-08 22-56-08](https://user-images.githubusercontent.com/47316824/89719615-6cc9b100-d9ca-11ea-9c0c-6ac1a24b2ccd.png)

`Unhandled Rejection (Error): The method eth_sendTransaction does not exist/is not available
send/<
src/index.ts:202
fulfilled
node_modules/tslib/tslib.es6.js:71
This screen is visible only in development. It will not appear if the app crashes in production.
Open your browser’s developer console to further inspect this error.  Click the 'X' or hit ESC to dismiss this message.`

Thanks

**Comment by pedrouid** [comment]
Could you try the following which we are currently testing?

1. Install pre-release version available on NPM at `next` tag (v1.2.0-alpha.0)
2. Make sure that you get the Signer from the provider

Example:
```js
const web3Provider = new WalletConnectProvider(..)

const provider = new ethers.providers.Web3Provider(web3Provider)

const tokenContract = new Contract(tokenAddress, tokenAbi, provider.getSigner())

```

**Comment by mds1** [comment]
@pedrouid Just gave that a shot and everything seemed to work as expected! Details:

- @walletconnect/web3-provider v1.2.0-alpha.0 
- web3modal v1.90
- Sent an ordinary ETH transaction with Argent

**Comment by pi0neerpat** [comment]
IT WORKS!!!!

```json
    "@ethersproject/contracts": "^5.0.1",
    "@ethersproject/providers": "^5.0.4",
    "@walletconnect/web3-provider": "1.2.0-alpha.0",
```

**Comment by pedrouid** [comment]
Fixed with [v1.2.0](https://github.com/WalletConnect/walletconnect-monorepo/releases/tag/1.2.0)

**Comment by KarthickSakthi** [comment]
> Could you try the following which we are currently testing?
> 
> 1. Install pre-release version available on NPM at `next` tag (v1.2.0-alpha.0)
> 2. Make sure that you get the Signer from the provider
> 
> Example:
> 
> ```js
> const web3Provider = new WalletConnectProvider(..)
> 
> const provider = new ethers.providers.Web3Provider(web3Provider)
> 
> const tokenContract = new Contract(tokenAddress, tokenAbi, provider.getSigner())
> ```

@pedrouid , I am using ether.js (v "5.6.9") and wagmi (v 0.6.1) for Connect wallet. I am using WalletConnectConnector and MetamaskConnector from wagmi. In Every Transaction Call from my dapp. I am following the below:

```
const provider = new ethers.providers.Web3Provider(window.ethereum);
    let signer = provider.getSigner();
    let signerAddress = await signer.getAddress();
```

So, when i connected with metamask the provider will work due to Window.ethereum!

But when I connected with WalletConnect via wagmi.. The account is connecting to my dapp, but for Transaction calls, it just call my metamask's account or getting metamask error .. due to (window.ethereum) provider . instead it needs to be reflected in my mobile to sign the transaction as per WalletConnect behaviour.

From the above convos I understand to connect WalletConnect Provider!... But for my Case, How do i analyse the Provider dynamically for When I connect with metamask, In all transactin call or providers should be for (window.ethereum)! and if Connected to my dapp via WalletConnect na , In all transactin call or providers should be for WalletConnect Provider.

Kindly Help on this!
anyone

@moodysalem @BSathvik @ricmoo @mds1
